<a href="https://colab.research.google.com/github/Abdelrhman410/DB/blob/main/Macro_Group.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Importing libraries and data

In [ ]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as mp
import seaborn as sb

In [ ]:
from google.colab import drive

drive.mount("/content/gdrive")

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
df = pd.read_excel("/content/gdrive/My Drive/XXCopy of UC- ORIGINAL MACRO SALES REPORT JAN JULY 2022.xlsx")

In [ ]:
df.head()

DISTRIBUTOR_NAME    Date  IQVIA_CUST_CODE                       NAME  \
0              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
1              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
2              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
3              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
4          IBNSINA  202208           100707  عبد الرحمن حسن عبد الرحمن   

                      Customer Name           STREET    DISTRICT   CITY  \
0  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
1  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
2  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
3  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
4  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   

  CMFTYPE MACRO_BRICKS_CODE  ... REP Territory Name District Name   Area Name  \
0    PHAR              B139  ...     M6_South Sohag  M6_U.E south  M6_UE Zone   
1    PHAR              B139  ...         OTC_UPE NA   OTC_COUNTRY    OTC AREA   
2    PHAR              B139  ...             VACANT        VACANT      VACANT   
3    PHAR              B139  ...           M7_Sohag        M7_U.E       M7_NA   
4    PHAR              B139  ...     M5_South Sohag  M5_U.E South  M5_UE Zone   

  MACRO_REGIONS              MACRO_PROD_DESC    BRAND     LINE SALES_UNITS  \
0   Upper Egypt      Frost Massage Gel 100gm    FROST  MACRO 6           1   
1   Upper Egypt             Primo plus 120ml    PRIMO      OTC           2   
2   Upper Egypt           Senzo-1 Soap 100gm    SENZO   OTHERS           2   
3   Upper Egypt  Synobar Skin cleanser 250ml  SYNOBAR  MACRO 7           1   
4   Upper Egypt               Scaro Gel 50gm    SCARO  MACRO 5           1   

  BONUS_UNITS  SALES_VALUES  
0           0         41.92  
1           0         50.30  
2           0         50.30  
3           0         69.86  
4           0        350.30  

[5 rows x 22 columns]

#Data Wrangling

###**Questions was given to me by the client and they are :**

1. Top  3 lines 

2. Lowest 3 lines 

3. Top 10 products

4. Lowest 10 products 

5. Top 5 district managers 

6. Top 5 medical rep

7. Top 5 cities 

8. Top 5 MACRO_BRICKS_NAME

9. Top 5 area name 

10. Top 3 brand 





##1. Check duplicates

In [ ]:
df.duplicated().sum()

0

**Note 1:** No duplicates

##2. Check coulmn types

In [ ]:
df.dtypes

DISTRIBUTOR_NAME       object
Date                    int64
IQVIA_CUST_CODE         int64
NAME                   object
Customer Name          object
STREET                 object
DISTRICT               object
CITY                   object
CMFTYPE                object
MACRO_BRICKS_CODE      object
MACRO_TERRITOTRIES     object
MACRO_BRICKS_NAME      object
REP Territory Name     object
District Name          object
Area Name              object
MACRO_REGIONS          object
MACRO_PROD_DESC        object
BRAND                  object
LINE                   object
SALES_UNITS             int64
BONUS_UNITS             int64
SALES_VALUES          float64
dtype: object

**Note 2:** Date coulmn needs to be change to date time format

##3. Check null values

In [ ]:
df.isnull().sum()

DISTRIBUTOR_NAME      0
Date                  0
IQVIA_CUST_CODE       0
NAME                  0
Customer Name         0
STREET                0
DISTRICT              0
CITY                  0
CMFTYPE               0
MACRO_BRICKS_CODE     0
MACRO_TERRITOTRIES    0
MACRO_BRICKS_NAME     0
REP Territory Name    0
District Name         0
Area Name             0
MACRO_REGIONS         0
MACRO_PROD_DESC       0
BRAND                 0
LINE                  0
SALES_UNITS           0
BONUS_UNITS           0
SALES_VALUES          0
dtype: int64

**Note 3:**  There is no null values

**Important Note:** I want to thank the data collector, and I hope that every data collector can be like him XDD.

In [ ]:
df.head(5)

DISTRIBUTOR_NAME    Date  IQVIA_CUST_CODE                       NAME  \
0              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
1              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
2              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
3              UCP  202208           100707  عبد الرحمن حسن عبد الرحمن   
4          IBNSINA  202208           100707  عبد الرحمن حسن عبد الرحمن   

                      Customer Name           STREET    DISTRICT   CITY  \
0  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
1  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
2  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
3  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   
4  عبد الرحمن حسن عبد الرحمن-100707  ش ميدان الشهداء  دار السلام  سوهاج   

  CMFTYPE MACRO_BRICKS_CODE  ... REP Territory Name District Name   Area Name  \
0    PHAR              B139  ...     M6_South Sohag  M6_U.E south  M6_UE Zone   
1    PHAR              B139  ...         OTC_UPE NA   OTC_COUNTRY    OTC AREA   
2    PHAR              B139  ...             VACANT        VACANT      VACANT   
3    PHAR              B139  ...           M7_Sohag        M7_U.E       M7_NA   
4    PHAR              B139  ...     M5_South Sohag  M5_U.E South  M5_UE Zone   

  MACRO_REGIONS              MACRO_PROD_DESC    BRAND     LINE SALES_UNITS  \
0   Upper Egypt      Frost Massage Gel 100gm    FROST  MACRO 6           1   
1   Upper Egypt             Primo plus 120ml    PRIMO      OTC           2   
2   Upper Egypt           Senzo-1 Soap 100gm    SENZO   OTHERS           2   
3   Upper Egypt  Synobar Skin cleanser 250ml  SYNOBAR  MACRO 7           1   
4   Upper Egypt               Scaro Gel 50gm    SCARO  MACRO 5           1   

  BONUS_UNITS  SALES_VALUES  
0           0         41.92  
1           0         50.30  
2           0         50.30  
3           0         69.86  
4           0        350.30  

[5 rows x 22 columns]

**Total Notes:**

1. No duplicates
2. Date coulmn needs to be change to date time format
3. There is no null valuesfds
4. Depending on the client questions I found that I will not use these coulmns " DISTRIBUTOR_NAME, Date, Iqvia cust code, name, Customer name, street, city, cmftype, macro bricks code, rep territory name, MACRO_TERRITOTRIES" so i will drop them

In [ ]:
df.drop(['DISTRIBUTOR_NAME','Date', 'IQVIA_CUST_CODE', 'NAME', 'Customer Name', 'STREET', 'DISTRICT', 'CMFTYPE', 'MACRO_BRICKS_CODE', 'MACRO_TERRITOTRIES'],  axis=1, inplace=True)

In [ ]:
df.head()

CITY MACRO_BRICKS_NAME REP Territory Name District Name   Area Name  \
0  سوهاج           Sohag 3     M6_South Sohag  M6_U.E south  M6_UE Zone   
1  سوهاج           Sohag 3         OTC_UPE NA   OTC_COUNTRY    OTC AREA   
2  سوهاج           Sohag 3             VACANT        VACANT      VACANT   
3  سوهاج           Sohag 3           M7_Sohag        M7_U.E       M7_NA   
4  سوهاج           Sohag 3     M5_South Sohag  M5_U.E South  M5_UE Zone   

  MACRO_REGIONS              MACRO_PROD_DESC    BRAND     LINE  SALES_UNITS  \
0   Upper Egypt      Frost Massage Gel 100gm    FROST  MACRO 6            1   
1   Upper Egypt             Primo plus 120ml    PRIMO      OTC            2   
2   Upper Egypt           Senzo-1 Soap 100gm    SENZO   OTHERS            2   
3   Upper Egypt  Synobar Skin cleanser 250ml  SYNOBAR  MACRO 7            1   
4   Upper Egypt               Scaro Gel 50gm    SCARO  MACRO 5            1   

   BONUS_UNITS  SALES_VALUES  
0            0         41.92  
1            0         50.30  
2            0         50.30  
3            0         69.86  
4            0        350.30

#Analyze Data

In [ ]:
df = df [df['LINE']!="غير معروف" ]

In [ ]:
Lines = {}

for Line in df["LINE"]:  
  if Line not in Lines:
    filtered_Sales = df[df['LINE']== Line ]["SALES_VALUES"].sum()
    Lines[Line] = filtered_Sales

In [ ]:
sort_data = sorted(Lines.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
print(sort_data_dict)

{'MACRO 1': 11921951.080000002, 'MACRO 2': 11388801.299999999, 'OTC': 9200660.693999998, 'MACRO 5': 8649952.91, 'MACRO 4': 7752972.189, 'MACRO 7': 7097652.1, 'MACRO 6': 5846212.47, 'MACRO 3': 5390783.95, 'OTHERS': 1548106.6199999999, 'VERDEX': 1497507.33, 'ATRAKTA': 1440565.6600000001, 'SOLODEX': 946265.7499999999}


In [ ]:
for i in range (3):
  print(list(sort_data_dict.items())[i])

('MACRO 1', 11921951.080000002)
('MACRO 2', 11388801.299999999)
('OTC', 9200660.693999998)


In [ ]:
for i in range (-3,0):
  print(list(sort_data_dict.items())[i])

('VERDEX', 1497507.33)
('ATRAKTA', 1440565.6600000001)
('SOLODEX', 946265.7499999999)


#**$\color{skyblue}{\text{Top 3 lines}}$**
"MACRO 1" => 11921951.08 revenue

"MACRO 2" => 11388801.3 revenue

"OTC" =>  9200660.69 revenue



---



 

#**$\color{skyblue}{\text{Lowest 3 lines}}$**
"SOLODEX"=> 946265.75 revenue

"ATRAKTA"=> 1440565.66 revenue

"VERDEX"=> 1497507.33 revenue

---



In [ ]:
print(df.columns.tolist())

['CITY', 'MACRO_BRICKS_NAME', 'REP Territory Name', 'District Name', 'Area Name', 'MACRO_REGIONS', 'MACRO_PROD_DESC', 'BRAND', 'LINE', 'SALES_UNITS', 'BONUS_UNITS', 'SALES_VALUES']


In [ ]:
Products = {}

for Product in df["MACRO_PROD_DESC"]:  
  if Product not in Products:
    filtered_Sales = df[df['MACRO_PROD_DESC']== Product ]["SALES_VALUES"].sum()
    Products[Product] = filtered_Sales

In [ ]:
sort_data = sorted(Products.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
print(sort_data_dict)

{'Orovex Peppermint 250ml': 5001567.81, 'Gold Plus Cleanser 250ml': 3937879.9700000007, 'Scaro Gel 50gm': 3346304.41, 'Frost Massage Spray 100ml': 2717577.08, 'Synobar Skin cleanser 250ml': 2679505.59, 'Orovex Clove 250ml': 2183120.55, 'Dermocalm Lotion 120ml': 2142031.97, 'Bringo Lotion 120ml': 2133319.46, 'Natry Soap 100gm': 1913797.6099999999, 'Care by Care Cream 100gm': 1847981.21, 'Vividol Hair Cream 100gm': 1815802.57, 'Macro Panthenol Cream 50gm': 1806471.65, 'Lit up Soap 100gm': 1665118.88, 'Frost Massage Gel 100gm': 1659670.6099999999, 'Revarest Plus Cleanser 250ml': 1588404.5599999998, 'Scaro Plus Cream 50gm': 1505269.8499999999, 'Mondo Cream 50gm': 1451558.98, 'Cordo Plus Spray 60 ml': 1362092.3190000001, 'Topi Gent Sebum Foam 150ml': 1332006.6099999999, 'Seroleze Hair Serum 120ml': 1288627.84, 'Klenva Cleanser 250ml': 1153406.32, 'Tritect Acne soap 100gm': 1094247.5599999998, 'Topi Gent Delicate Foam 150ml': 1086723.3900000001, 'Synobar Soap 100 gm': 1016132.26, 'Marvel Cre

In [ ]:
for i in range (10):
  print(list(sort_data_dict.items())[i])

('Orovex Peppermint 250ml', 5001567.81)
('Gold Plus Cleanser 250ml', 3937879.9700000007)
('Scaro Gel 50gm', 3346304.41)
('Frost Massage Spray 100ml', 2717577.08)
('Synobar Skin cleanser 250ml', 2679505.59)
('Orovex Clove 250ml', 2183120.55)
('Dermocalm Lotion 120ml', 2142031.97)
('Bringo Lotion 120ml', 2133319.46)
('Natry Soap 100gm', 1913797.6099999999)
('Care by Care Cream 100gm', 1847981.21)


In [ ]:
prod = [value for value in list(sort_data_dict.items()) if value[1] >0]

In [ ]:
prod[-10:][::-1]

[('Verdex - Eye Contour Aid 50ml', 195.62),
 ('Verdex - Alba cream 50ml', 201.2),
 ('VAKTER 60 ML D', 224.2),
 ('Verdex - Sun Off Junior 50gm', 271.4),
 ('Verdex - Hair Forte Cream 100ml', 294.96000000000004),
 ('Cordo Hand Gel 60ml', 399.13),
 ('Verdex - Hair Forte Shampoo 200ml', 425.65999999999997),
 ('Octenidine Oral spray (120ml)', 504.4),
 ('Verdex - Alba serum 50ml', 643.8100000000001),
 ('Verdex - Sypo Clear 50ml', 715.4)]

#**$\color{skyblue}{\text{Top 10 Products}}$**


"Orovex Peppermint 250ml" => 5001567.81 revenue

"Gold Plus Cleanser 250ml" =>  3937879.97 revenue

"Scaro Gel 50gm" =>  3346304.41
 revenue

"Frost Massage Spray 100ml" =>   2717577.08 revenue

"Synobar Skin cleanser 250ml" =>   2679505.59 revenue

"Orovex Clove 250ml"=> 2183120.55  revenue


"Dermocalm Lotion 120ml"=> 2142031.97  revenue


"Bringo Lotion 120ml"=> 2133319.46  revenue


"Natry Soap 100gm"=> 1913797.61  revenue


"are by Care Cream 100gm"=> 1847981.21  revenue









#**$\color{skyblue}{\text{Lowest 10 Products}}$**


"Verdex - Eye Contour Aid 50ml" => 195.62

 "Verdex - Alba cream 50ml" => 201.2
 
 "VAKTER 60 ML D" => 224.2
 
 "Verdex - Sun Off Junior 50gm" => 271.4
 
 "Verdex - Hair Forte Cream 100ml" => 294.96
 
 "Cordo Hand Gel 60ml" => 399.13
 
 "Verdex - Hair Forte Shampoo 200ml" => 425.659
 
 "Octenidine Oral spray (120ml)" => 504.4
 
 "Verdex - Alba serum 50ml" => 643.81
 
 "Verdex - Sypo Clear 50ml" => 715.4

In [ ]:
Districts_Names = {}

for District in df["District Name"]:  
  if District not in Districts_Names:
    filtered_Sales = df[df['District Name']== District ]["SALES_VALUES"].sum()
    Districts_Names[District] = filtered_Sales

In [ ]:
sort_data = sorted(Districts_Names.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
list(sort_data_dict.items())[:5]

[('M7_Delta 1', 3669948.5300000003),
 ('M2_Cairo East', 3389894.07),
 ('OTC_DM Delta', 3275675.06),
 ('M2_Alex Zone', 3167995.3399999994),
 ('M5_Delta West', 2766359.4199999995)]


#**$\color{skyblue}{\text{Top 5 District}}$**




"M7_Delta 1" => 3669948.53

 "M2_Cairo East" => 3389894.07

 "OTC_DM Delta" => 3275675.06

 "M2_Alex Zone" => 3167995.34
 
 "M5_Delta West" => 2766359.42

In [ ]:
Medical_rep = {}

for rep in df["REP Territory Name"].unique():  
  filtered_Sales = df[df['REP Territory Name']== rep ]["SALES_VALUES"].sum()
  Medical_rep[rep] = filtered_Sales

sort_data = sorted(Medical_rep.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
list(sort_data_dict.items())[:5]

[('M2_Heliopolis - Zaytoun', 3174139.93),
 ('M2_Behara South', 2827718.9499999997),
 ('M7_Gharbia 1', 2076710.23),
 ('M1_Dakahlia 1', 1941026.57),
 ('M3_Gharbia 1', 1863889.09)]


#**$\color{skyblue}{\text{Top 5 Medical Rep}}$**




"M2_Heliopolis - Zaytoun => 3174139.93

 "M2_Behara South => 2827718.95
 
 "M7_Gharbia 1 => 2076710.23
 
 "M1_Dakahlia 1 => 1941026.57
 
 "M3_Gharbia 1 => 1863889.09

In [ ]:
Cities = {}

for City in df["CITY"].unique():  
  filtered_Sales = df[df['CITY']== City ]["SALES_VALUES"].sum()
  Cities[City] = filtered_Sales

sort_data = sorted(Cities.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
list(sort_data_dict.items())[:5]

[('القاهره', 15347563.058000002),
 ('الدقهليه', 12836442.09),
 ('الغربيه', 12358264.177),
 ('الجيزه', 7852025.577),
 ('البحيره', 4390424.079999999)]

#**$\color{skyblue}{\text{Top 5 Cities}}$**

القاهره" 15347563.058000002"

 الدقهليه" 12836442.09"

 الغربيه" 12358264.177"

 الجيزه" 7852025.577"
 
 البحيره" 4390424.079999999

In [ ]:
MACRO_BRICKS_NAME = {}

for BRICKS in df["MACRO_BRICKS_NAME"].unique():  
  filtered_Sales = df[df['MACRO_BRICKS_NAME']== BRICKS ]["SALES_VALUES"].sum()
  MACRO_BRICKS_NAME[BRICKS] = filtered_Sales

sort_data = sorted(MACRO_BRICKS_NAME.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
list(sort_data_dict.items())[:5]

[('Gharbia 1', 9890581.200000001),
 ('Dakahlia 3', 6289635.580000001),
 ('Heliopolis 4', 5802062.45),
 ('Dakahlia 1', 5008817.45),
 ('Heliopolis 3', 3395234.6299999994)]

#**$\color{skyblue}{\text{Top 5 Macro Bricks}}$**

"Gharbia 1" => 9890581.2

 "Dakahlia 3" => 6289635.58

 "Heliopolis" 4 => 5802062.45

 "Dakahlia 1" => 5008817.45

 "Heliopolis 3" => 3395234.6

In [ ]:
area_names = {}

for area in df["Area Name"].unique():  
  filtered_Sales = df[df['Area Name']== area ]["SALES_VALUES"].sum()
  area_names[area] = filtered_Sales

sort_data = sorted(area_names.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
list(sort_data_dict.items())[:5]

[('OTC AREA', 9200588.695),
 ('M7_NA', 7097652.1),
 ('M1_Delta Zone', 5779142.340000001),
 ('M1_Cairo Zone', 4319688.417),
 ('M2_Cairo Zone', 4173713.34)]

#**$\color{skyblue}{\text{Top 5 Areas}}$**

"OTC AREA" => 9200588.695

"M7_NA" => 7097652.1

"M1_Delta Zone" => 5779142.34

"M1_Cairo Zone" => 4319688.417

"M2_Cairo Zone" => 4173713.34

In [ ]:
brands = {}

for brand in df["BRAND"].unique():  
  filtered_Sales = df[df['BRAND']== brand ]["SALES_VALUES"].sum()
  brands[brand] = filtered_Sales

sort_data = sorted(brands.items(), key=lambda x: x[1], reverse=True)
sort_data_dict = dict(sort_data)
list(sort_data_dict.items())[:3]

[('OROVEX', 11422186.360000001),
 ('SCARO', 4927506.43),
 ('FROST', 4557913.109999999)]

#**$\color{skyblue}{\text{Top 5 Brands}}$**

"OROVEX => 11422186.36

 "SCARO => 4927506.43
 
 "FROST => 4557913.11